### PR TITLE
GPGS Authorization and Saved games functionality using JNI

### PR DESCRIPTION
**Pull request by AGulev**

### CODE REVIEW

File: gpgs/src/gpgs_extension.cpp [hunk 1]
```diff
@@ -1,343 +1,640 @@
...
 };
 
-struct Gpg
+struct GPGS_Disk
 {
-    Gpg() : m_AuthInProgress(false) {}
-    bool m_AuthInProgress;
-    LuaCallbackInfo m_Callback;
-    std::unique_ptr<gpg::GameServices> m_GameServices;
+    bool                   is_using;
```
Comment (jhonnyking) [inline gpgs/src/gpgs_extension.cpp]: why not m_IsUsing?

File: gpgs/src/gpgs_jni.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	}
+	jclass load(const char *class_name) {
+		jstring str_class_name = env->NewStringUTF(class_name);
+		jclass loaded_class = (jclass)env->CallObjectMethod(class_loader_object, find_class, str_class_name);
+		env->DeleteLocalRef(str_class_name);
+		return loaded_class;
+	}
+};
+
+#endif
```
Comment (jhonnyking) [inline gpgs/src/gpgs_jni.h]: newline

File: gpgs/src/gpgs_extension.cpp [hunk 3]
```diff
@@ -1,343 +1,640 @@
...
+dmExtension::Result FinalizeGpg(dmExtension::Params* params)
 {
     return dmExtension::RESULT_OK;
 }
 
-DM_DECLARE_EXTENSION(LIB_NAME, "Gpg", AppInitializeGpg, AppFinalizeGpg, 0, 0, 0, 0)
+DM_DECLARE_EXTENSION(EXTENSION_NAME, LIB_NAME, 0, 0, InitializeGpg, 0, 0, FinalizeGpg)
 
-#endif
+#endif
```
Comment (jhonnyking) [inline gpgs/src/gpgs_extension.cpp]: newline

File: gpgs/src/java/com/defold/gpgs/GpgsJNI.java [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+
+    //--------------------------------------------------
+    // GoogleDrive (Snapshots)
+
+    // Client used to interact with Google Snapshots.
+    private SnapshotsClient mPlayerSnapshotsClient = null;
+    private Snapshot mPlayerSnapshot = null;
+    private byte[] currentplayerSave = null;
+
+    private int maxCoverImageSize = 819200;
```
Comment (jhonnyking) [inline gpgs/src/java/com/defold/gpgs/GpgsJNI.java]: Where did you get these values from? Is it from the official docs?

File: gpgs/src/private_gpgs_callback.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+{
+    MESSAGE_ID msg;
+    char* json;
+};
+
+void gpgs_set_callback(lua_State* L, int pos);
+void gpgs_callback_initialize();
+void gpgs_callback_finalize();
+void gpgs_callback_update();
+void gpgs_add_to_queue(MESSAGE_ID msg, const char*json);
```
Comment (jhonnyking) [inline gpgs/src/private_gpgs_callback.h]: newline

File: gpgs/src/utils/LuaUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
+#include "LuaUtils.h"
+
+bool luaL_checkbool(lua_State *L, int numArg)
+{
+  bool b = false;
```
Comment (jhonnyking) [inline gpgs/src/utils/LuaUtils.cpp]: Indentation in this file is different, is this our/your code or someone elses?

File: gpgs/src/gpgs_jni.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
+#if defined(DM_PLATFORM_ANDROID)
```
Comment (jhonnyking) [inline gpgs/src/gpgs_jni.h]: No header guard?

File: gpgs/src/java/com/defold/gpgs/GpgsJNI.java (hunk 4 above)
Comment (AGulev) [reply gpgs/src/java/com/defold/gpgs/GpgsJNI.java]: Yes, I'll add a comment

File: gpgs/src/utils/LuaUtils.cpp (hunk 6 above)
Comment (AGulev) [reply gpgs/src/utils/LuaUtils.cpp]: Yes, this is from my DefVideoAds plugin. I'll fix it, thx!

File: gpgs/src/com_defold_gpgs_GpgsJNI.h [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#include <jni.h>
+/* Header for class com_defold_gpgs_GpgsJNI */
+
+#ifndef _Included_com_defold_gpgs_GpgsJNI
+#define _Included_com_defold_gpgs_GpgsJNI
+#ifdef __cplusplus
+extern "C" {
+	#endif
```
[the remark this answers is not in the document]
Comment (AGulev) [reply gpgs/src/com_defold_gpgs_GpgsJNI.h]: Thx

**Comment by AGulev** [comment]
@jhonnyking fixed. Thanks for the review!

**Pull request merged**